### PR TITLE
feat: restore backup in local container

### LINKED ISSUE #32: Feature: restore in a local container for dev purpose
Hey everyone 👋🏽  I have an idea that can make developer life so much easier while developing locally. What do you think of creating a command like this one 👇

```
$ replibyte restore -v latest --local
Restore complete!

To connect to your Postgres database, use the following connection string:
DATABASE_URL=postgres://root:password@localhost:5432/root
```

This command would start a local docker of the appropriate database (E.g Postgres) and restore the latest dump into it.

## Main benefits

Work locally with safe production data.

Happy to have your feedback.

### PR DESCRIPTION
**Pull request by fabriceclementz**
This PR add support for restoring a backup inside a Docker container as discussed in #32.

> This is currently a wip, needs some tests and improvements

Closes #32

### CODE REVIEW

File: replibyte/src/destination/postgres_docker.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        PostgresDocker {}
+    }
+}
+
+impl Connector for PostgresDocker {
+    fn init(&mut self) -> Result<(), Error> {
+        let _ = binary_exists("docker")?;
+
+        // Start the container and configure the network and credentials to be able to connect locally.
+        // TODO: use free port, handle postgres version and container name.
```
Comment (evoxmusic) [inline replibyte/src/destination/postgres_docker.rs:21]: maybe the port can be also configurable by the user

File: replibyte/src/main.rs [hunk 2]
```diff
@@ -290,6 +292,15 @@ fn main() -> anyhow::Result<()> {
                     },
                 };
 
+                if cmd.local {
+                    // TODO: handle Postgres and MongoDB
+                    let task = FullRestoreTask::new(PostgresDocker::default(), bridge, options);
+                    let _ = task.run(progress_callback)?;
+
+                    // TODO: Return the credentials to the user.
```
Comment (evoxmusic) [inline replibyte/src/main.rs:300]: We'll also need to keep the container running until the user press CTRL/CMD + C (Sigterm) that we need to handle to drop the container.

File: replibyte/src/destination/postgres_docker.rs (hunk 1 above)
Comment (fabriceclementz) [reply replibyte/src/destination/postgres_docker.rs]: 👍 I've added an optional `--port` flag

File: replibyte/src/main.rs (hunk 2 above)
Comment (fabriceclementz) [reply replibyte/src/main.rs]: As a default, the container is stopped when the user press `ctrl+c` and `--remove` option was added to definitely remove the container.

**Comment by fabriceclementz** [comment]
@evoxmusic @benny-n `mongorestore` inside a docker container seems a bit tricky, I have a corruption error, maybe you can take a look ? despite that, the dump seems restored correctly.

```
    Running `target/debug/replibyte -c ./examples/source-and-dest-mongodb-bridge-minio.yaml restore local --value latest --image mongodb --port 27020`
2022-04-09T13:26:00.621+0000	preparing collections to restore from
2022-04-09T13:26:00.635+0000	reading metadata for test.cars from archive on stdin
2022-04-09T13:26:00.635+0000	reading metadata for test.states from archive on stdin
2022-04-09T13:26:00.636+0000	reading metadata for test.users from archive on stdin
2022-04-09T13:26:00.655+0000	restoring test.cars from archive on stdin
2022-04-09T13:26:00.668+0000	finished restoring test.cars (10 documents, 0 failures)
2022-04-09T13:26:00.690+0000	restoring test.states from archive on stdin
2022-04-09T13:26:00.717+0000	restoring test.users from archive on stdin
2022-04-09T13:26:00.734+0000	finished restoring test.users (10 documents, 0 failures)
2022-04-09T13:26:00.736+0000	finished restoring test.states (10 documents, 0 failures)
2022-04-09T13:26:00.736+0000	restoring users from archive on stdin
2022-04-09T13:26:00.781+0000	no indexes to restore for collection test.states
2022-04-09T13:26:00.781+0000	no indexes to restore for collection test.users
2022-04-09T13:26:00.781+0000	no indexes to restore for collection test.cars
2022-04-09T13:26:00.782+0000	Failed: corruption found in archive; I/O error reading length or terminator ( unexpected EOF )
2022-04-09T13:26:00.782+0000	30 document(s) restored successfully. 0 document(s) failed to restore.
Error: command error: "exit status: 1"
```

**Comment by evoxmusic** [comment]
I am on the Database subset - But I can take a look when it'll be done (hopefully this weekend)

**Comment by fabriceclementz** [comment]
thx @evoxmusic FYI, I've updated the CLI for the restore command as :

Local restore : ` replibyte restore local [OPTIONS] --value <[latest | backup name]> --image <[postgresql | mysql | mongodb]>` with the following options

```sh
 -p, --port <PORT>
            Docker container port to map on the host

    -r, --remove
            Remove the Docker container on Ctrl-c

    -t, --tag <TAG>
            Docker image tag for the container to spawn
```
maybe the image flag could be deleted if we store the database type (eg: mongodb, postgresql, mysql) inside the backup, this way we could determinate the container to use for restoring data.

---

and remote restore which correspond to the configured destination in the config file

```sh
replibyte restore remote [OPTIONS] --value <[latest | backup name]>
```

the options remain unchanged in this case.

The reason about this separation is that the restore local options (eg: port, tag, remove) didn't make sense to me when we want to restore in the configured destination in the config file. Let me know what you think when you have some time :)

**Comment by evoxmusic** [comment]
I agree with all your suggestions. For the local restore we should remove the image flag and store the database information with the backup. I even think we should store almost all the backup options (but this can be done into a separate PR). 

I can't wait to give it a try :)

**Comment by benny-n** [comment]
> @evoxmusic @benny-n `mongorestore` inside a docker container seems a bit tricky, I have a corruption error, maybe you can take a look ? despite that, the dump seems restored correctly.
> 
> ```
>     Running `target/debug/replibyte -c ./examples/source-and-dest-mongodb-bridge-minio.yaml restore local --value latest --image mongodb --port 27020`
> 2022-04-09T13:26:00.621+0000	preparing collections to restore from
> 2022-04-09T13:26:00.635+0000	reading metadata for test.cars from archive on stdin
> 2022-04-09T13:26:00.635+0000	reading metadata for test.states from archive on stdin
> 2022-04-09T13:26:00.636+0000	reading metadata for test.users from archive on stdin
> 2022-04-09T13:26:00.655+0000	restoring test.cars from archive on stdin
> 2022-04-09T13:26:00.668+0000	finished restoring test.cars (10 documents, 0 failures)
> 2022-04-09T13:26:00.690+0000	restoring test.states from archive on stdin
> 2022-04-09T13:26:00.717+0000	restoring test.users from archive on stdin
> 2022-04-09T13:26:00.734+0000	finished restoring test.users (10 documents, 0 failures)
> 2022-04-09T13:26:00.736+0000	finished restoring test.states (10 documents, 0 failures)
> 2022-04-09T13:26:00.736+0000	restoring users from archive on stdin
> 2022-04-09T13:26:00.781+0000	no indexes to restore for collection test.states
> 2022-04-09T13:26:00.781+0000	no indexes to restore for collection test.users
> 2022-04-09T13:26:00.781+0000	no indexes to restore for collection test.cars
> 2022-04-09T13:26:00.782+0000	Failed: corruption found in archive; I/O error reading length or terminator ( unexpected EOF )
> 2022-04-09T13:26:00.782+0000	30 document(s) restored successfully. 0 document(s) failed to restore.
> Error: command error: "exit status: 1"
> ```

@fabriceclementz 
I think I know what this is about.. 
When I implemented the mongo destination I noticed that I'm getting the archive input as a null-terminated string (the last byte was always `0x00`).
Since the mongorestore parser always expects to find 4 bytes of `0xFF` at the end of each archive, I had to [manually remove it](https://github.com/Qovery/replibyte/blob/main/replibyte/src/destination/mongodb.rs#L76).
Perhaps we should add a safety check to see if the archive string is null-terminated or not, and only remove the last byte if it is indeed the null byte.

**Comment by fabriceclementz** [comment]
Thanks @benny-n , It works ! 
The last bytes seems to end with 

```
    0xFF,
    0xFF,
    0xFF,
    0xFF,
    0x0A,
```

So for now, I've only removed the last byte like you did in the mongodb destination. I haven't added a safety check.

**Comment by fabriceclementz** [comment]
Hi @evoxmusic 
No, I will rebase the branch tonight and I think it will be good enough for merge 🙂

**Comment by fabriceclementz** [comment]
@evoxmusic It's ready to review / merge.

As we discuss here https://github.com/Qovery/replibyte/pull/39#issuecomment-1094124652 , I will submit another PR to remove the image flag by adding the DB type in the backup.

**Comment by evoxmusic** [comment]
Well done Fabrice. I'll take a look tomorrow ☺️

**Review by evoxmusic — APPROVED** [review]
You did an amazing job @fabriceclementz - I have nothing to add. In the future, we'll just need to be able to load locally the right database version. We'll need to store this information in the manifest file with the backups. Then from that, we'll be able to use the appropriate MongoDB, Postgres, or MySQL container version.

**Pull request merged**